### ISSUE TITLE
Configure should be warning-free

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#8132]
Configure should have 'use warnings' added to it, and all warnings fixed up.

In fact, all perl scripts should have 'use strict; use warnings;' in them.  One way to get that list is:
```
; git ls-files | grep '\.p[lm]' | xargs grep -L 'use strict'
```

**Comment by mspncp** [comment]
What about those locations? Should they be adopted, too? Or should they remain unchanged?

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L498

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L536

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L549

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L550

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L1479

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L1487

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L1492

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L1508

https://github.com/openssl/openssl/blob/62b563b9df161a992fde18a0cb0d1a0969158412/external/perl/Text-Template-1.46/lib/Text/Template.pm#L1524

**Comment by mspncp** [comment]
For anybody who dares to try it:

```
git ls-files | grep '\.p[lm]' | xargs sed -i 's/use strict;/use strict; use warnings;/g'
```

**Comment by mspncp** [comment]
Most of the keeps keep repeating, though:

```
$ /usr/bin/perl ./Configure debug-linux-x86_64 enable-crypto-mdebug shared --strict-warnings --prefix=/opt/openssl-dev
Configuring OpenSSL version 3.0.0-dev for target debug-linux-x86_64
Using os-specific seed configuration
Use of uninitialized value $files in split at /home/msp/openssl/openssl/test/build.info line 7.
Creating configdata.pm
Creating Makefile
Use of uninitialized value $Text::Template::GEN0::config{"RCFLAGS"} in array dereference at template line 222.
Use of uninitialized value $Text::Template::GEN0::config{"lib_cppflags"} in array dereference at template line 267.
Use of uninitialized value $Text::Template::GEN0::config{"shared_cppflag"} in array dereference at template line 267.
Use of uninitialized value $Text::Template::GEN0::config{"lib_cflags"} in array dereference at template line 276.
Use of uninitialized value $Text::Template::GEN0::config{"shared_cflag"} in array dereference at template line 276.
Use of uninitialized value $Text::Template::GEN0::config{"lib_cxxflags"} in array dereference at template line 281.
Use of uninitialized value $Text::Template::GEN0::config{"shared_cxxflag"} in array dereference at template line 281.
Use of uninitialized value $Text::Template::GEN0::config{"dso_cppflags"} in array dereference at template line 290.
Use of uninitialized value $Text::Template::GEN0::config{"module_cppflags"} in array dereference at template line 290.
Use of uninitialized value $Text::Template::GEN0::config{"dso_cflags"} in array dereference at template line 295.
Use of uninitialized value $Text::Template::GEN0::config{"module_cflags"} in array dereference at template line 295.
Use of uninitialized value $Text::Template::GEN0::config{"dso_cxxflags"} in array dereference at template line 300.
Use of uninitialized value $Text::Template::GEN0::config{"module_cxxflag"} in array dereference at template line 300.
Use of uninitialized value $Text::Template::GEN0::config{"dso_ldflags"} in array dereference at template line 305.
Use of uninitialized value $Text::Template::GEN0::config{"module_ldflags"} in array dereference at template line 305.
Use of uninitialized value $Text::Template::GEN0::config{"bin_cppflags"} in array dereference at template line 309.
Use of uninitialized value $Text::Template::GEN0::config{"bin_cflags"} in array dereference at template line 312.
Use of uninitialized value $Text::Template::GEN0::config{"bin_cxxflags"} in array dereference at template line 315.
Use of uninitialized value $Text::Template::GEN0::config{"bin_lflags"} in array dereference at template line 318.
Scalar value @args{"deps"} better written as $args{"deps"} at template line 1116.
Scalar value @args{"deps"} better written as $args{"deps"} at template line 1179.
Scalar value @args{"deps"} better written as $args{"deps"} at template line 1221.
Use of uninitialized value $obj in hash element at template line 1374.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1375.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $obj in hash element at template line 1376.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1377.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $args{"generator_deps"} in array dereference at template line 973.
Use of uninitialized value $obj in hash element at template line 1374.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1375.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $obj in hash element at template line 1376.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1377.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $args{"generator_deps"} in array dereference at template line 973.
Use of uninitialized value $obj in hash element at template line 1374.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1375.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $obj in hash element at template line 1376.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1377.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $args{"generator_deps"} in array dereference at template line 973.
Use of uninitialized value $obj in hash element at template line 1374.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1375.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $obj in hash element at template line 1376.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $bin in hash element at template line 1377.
Use of uninitialized value in array dereference at template line 1377.
Use of uninitialized value $args{"generator_deps"} in array dereference at template line 973.
Use of uninitialized value in array dereference at template line 1401.
Use of uninitialized value in array dereference at template line 1401.
Use of uninitialized value in array dereference at template line 1401.

...

```

**Comment by richsalz** [comment]
The template stuff is "external" and I am fine if we just leave it alone.  OUR CODE should be fixed.

**Comment by mspncp** [comment]
> The template stuff is "external" and I am fine if we just leave it alone. OUR CODE should be fixed.

Hmmm... after reverting the `external` directory, I don't get any warnings anymore.

Also, I just noticed that most of our perl modules already have a `use warnings;` declaration (on a separate line).

Does this mean there is nothing left to do, or am I missing something?


```
git ls-files | grep '\.p[lm]' | xargs sed -i 's/use strict;/use strict; use warnings;/g'
git checkout -- external
```

```
perl Configure no-idea no-camellia threads -fPIC --prefix=/home/msp/build/openssl shared debug-linux-x86_64
Configuring OpenSSL version 3.0.0-dev for target debug-linux-x86_64
Using os-specific seed configuration
Creating configdata.pm
Creating Makefile

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL file first)         ***
***                                                                ***
**********************************************************************
```

**Comment by richsalz** [comment]
Not everything has "use strict" or "use warnings"  The command-line I posted will lost those files without strict.  Make the obvious change to chck for "use warnings" and then fix the warnings.  Try
```
perl -cw ./Configure
```
for example.

There's lots of stuff to do.  Even just putting the "use" lines into the perl scripts that we wrote.

perl -cw Configure
```
for example.

**Comment by mspncp** [comment]
> The command-line I posted will list those files without strict.

Oh, thanks for the hint. I overlooked that you used a capital `-L`, not a `-l` option.

**Comment by levitte** [comment]
I agree with you, @richsalz, and am willing to put in some time to do this

**Comment by levitte** [comment]
I went with `git ls-files | grep '\.p[lm]' | while read F; do if grep -q -E 'use +strict' $F && grep -q -E 'use +warnings' $F; then :; else echo $F; fi; done`

That's quite a number of files...

**Comment by levitte** [comment]
Hmmm...  you know, I'm gonna avoid the perlasm files.  It's simply not worth it.
(they are very much Perl 4)

**Comment by mspncp** [comment]
> That's quite a number of files...

I could help you with the routine work: I put all the missing `use strict` and `use warnings` statements in and send you the result as a patch, then you 'only' have to fix the warnings. :-)

**Comment by richsalz** [comment]
#10965 fixes Configure.

**Comment by nhorman** [comment]
seems like based on the conversation above, this was attempted, and the amount of effort for the reward was deemed not worthwhile.

Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by richsalz** [comment]
I don't care.  If someone on the OpenSSL team does, feel free to re-open this.
